### PR TITLE
fix rerun before/after hooks on top navigation

### LINKED ISSUE #1987: Code in hooks and test are rerun when changing to a different superdomain
### Current behavior:

When running my spec file, I noticed that some code in the `before` hook (not `beforeEach`) is being run before each of my three test cases, causing my checks to fail. Want to remark that I am using different superdomains for every test, I don't see this when I run all the tests with the same superdomain.

### Desired behavior:

Code in the `before` hook only runs once, before all the tests, no matter if using different superdomains.

### Steps to reproduce:

I explain my setup. My spec file consists of:
```
- before()
- after()
- it() //superdomain 1
- it() //superdomain 2 using an iframe
- it() //superdomain 1
```
As I am using different superdomains, the only way that I've read to do it is visiting the different domains in different tests (https://docs.cypress.io/guides/guides/web-security.html#One-Superdomain-per-Test)

I've also set `"chromeWebSecurity": false`

Within the before hook I run several requests to some endpoints as a way to get some data that I will use through the tests; Get it once, and use it across the three it().

I've also set a `console.log('This only happens in the beforeHook:  ' + var)` to see what's the value of my variable, and this is what I see in the run:

- Running the first it()

 
- Second it()

- Third it()

After every `it()` the console gets cleared, which is fine, but I don't expect to see the log again during the second and the third tests, as this is supposed to run only once. Notice how the variable gets an increment for each loop, as I would expect had I done it in a `beforeEach` hook.

In the left side of the window every seems fine as there is no command logging for the `before` hook when running the second and third tests, but the `console.log` reveals that.

### Versions

- Cypress version: 3.0.1
- Browser: Chrome 66
- OS: Ubuntu Linux 16.04 LTS

### LINKED ISSUE #2296: Exception in root level after() doesn't fail test run in interactive mode
### Current behavior:

If there's an `after()` outside of any `describe`, `it` or `context` ("rool level"), which throws an exception, that doesn't fail the test run. This is different from how it works in `mocha`, where you get a `"after all" hook` failure.

Also, if it's a synchronous failure, the exception doesn't show up anywhere.
If it's async, it shows up like this:

screen shot 2018-08-07 at 5 13 10 pm

### Desired behavior:

The whole test run should fail (sometimes you would want to place an assertion there that is possible only after all tests have run).

### Steps to reproduce:

Add this to your tests:
```js
after(() => {
  cy.then(() => {
    throw new Error('bla bla');
  });
});
```

### Versions

Cypress: 3.0.3
OS: Mac OS X 10.12.6 (Sierra)
Browser: both Electron (59) and Chrome (68)

### LINKED ISSUE #7021: after/before hooks are rerun on domain navigation
### Current behavior:
on domain navigation, before/after hooks will be rerun erroneously.

### Desired behavior:
before/after hooks should only be run once regardless of domain navigation

### Test code to reproduce

this spec will infinitely navigate between 2 domains since the before and after hooks are rerun
```js
describe('suite', () => {
  before(() => {
    // will cause infinite top navigation
    cy.visit('/fixtures/generic.html')
  })

  beforeEach(() => {
  })

  it('test', () => {
  })

  afterEach(() => {
  })

  after(() => {
    // will cause infinite top navigation
    cy.visit('/fixtures/generic.html')
  })
})

describe('cause top navigation', () => {
  it('test', () => {
    cy.visit('http://localhost:3501/fixtures/generic.html')
  })
})

```

### Versions
all

### PR DESCRIPTION
**Pull request by kuceb**
- fix #1987
- fix #2296 failure in root after hook doesn't show failure in browser reporter
- address #2777 (fixes half of the problem - where hooks are rerun on navigation)

### User facing changelog
- Bug fix: We no longer rerun before/after hooks of already ran suites on domain navigation

### PR Tasks
- [x] rename duplicate issue number
- [x] add testcase for navigation inside suite
- [x] this introduced another bug due to other broken behavior in cypress related to global hooks, so this currently skips global after hooks after navigation - add small exclusion for this edge case

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub?

### CODE REVIEW

File: packages/driver/src/cypress/runner.js [hunk 1]
```diff
@@ -889,6 +889,16 @@ const create = function (specWindow, mocha, Cypress, cy) {
     return _testsById[id]
   }
 
+  function hasTestAlreadyRun (test) {
+    if (Cypress._RESUMED_AT_TEST) {
+      if (+test.id.slice(1) < +Cypress._RESUMED_AT_TEST.slice(1)) {
```
Comment (jennifer-shehane) [inline packages/driver/src/cypress/runner.js:894]: Maybe there's a clearer variable name that could be used here to describe this check. ID is a string here?

File: packages/driver/src/cypress/runner.js [hunk 2]
```diff
@@ -962,6 +972,10 @@ const create = function (specWindow, mocha, Cypress, cy) {
           break
       }
 
+      if (hasTestAlreadyRun(test)) {
+        return args[0]()
```
Comment (jennifer-shehane) [inline packages/driver/src/cypress/runner.js:976]: what is `args[0]()` also? Have to look into another file to even know what this args is and even then gives no indication on what the first args is.

File: packages/server/test/support/fixtures/projects/e2e/cypress/integration/issue_1987_spec.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+describe('suite', () => {
+  before(() => {
+    // will cause infinite top navigation
+    cy.visit('http://localhost:3434')
```
Comment (jennifer-shehane) [inline packages/server/test/support/fixtures/projects/e2e/cypress/integration/issue_1987_spec.js:4]: I don't understand how this test could ever work. This will be considered visiting domains of 2 different origins in a single test. What am I missing?

Screen Shot 2020-04-17 at 3 01 25 PM

File: packages/server/test/e2e/3_issue_1987.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      static: true,
+    },
+    {
+      port: 4545,
+      static: true,
+    }],
+  })
+
+  // https://github.com/cypress-io/cypress/issues/1987
+  // before/after hooks should not be rerun on top navigation
```
Comment (jennifer-shehane) [inline packages/server/test/e2e/3_issue_1987.ts:16]: you mention after hooks, but these aren't really tested in the test

```suggestion
  // before hooks should not be rerun on top navigation
```

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
This does address the issue of `before` hooks running multiple times when navigating to a different cross-domain across separate tests. (I actually wasn't able to replicate this happening with `after` hooks before this fix).

This does not address #2777 which is the `before` hook running twice when there is no `baseUrl`, so it navigates to our `localhost:randomport` - runs `before` hook, then redirects to the url under `cy.visit()` and runs the `before` hook again. Not sure if that is important for you to address to move forward with your other work.

File: packages/server/test/support/fixtures/projects/e2e/cypress/integration/issue_1987_spec.js (hunk 3 above)
Comment (Bkucera) [reply packages/server/test/support/fixtures/projects/e2e/cypress/integration/issue_1987_spec.js:4]: look at the order of the tests. The before hook is tied to an empty test, so the next test only has one navigation

Comment (jennifer-shehane) [reply packages/server/test/support/fixtures/projects/e2e/cypress/integration/issue_1987_spec.js:4]: tricky 👍

File: packages/driver/src/cypress/runner.js (hunk 1 above)
Comment (brian-mann) [inline packages/driver/src/cypress/runner.js:894]: i would add an additional `getId` function since thats what the `slice(1)` is doing, but its completely masked by this implementation. it looks arbitrary otherwise.

Comment (brian-mann) [inline packages/driver/src/cypress/runner.js:894]: you can also just `return (expression)` here instead of explicitly `return true` after the `if(...)`

**Review by brian-mann — CHANGES_REQUESTED** [review]

**Comment by Bkucera** [comment]
Well, I clicked the update branch button and it looks like it merged it instead

**Comment by chrisbreiding** [comment]
Does this also fix #3904?

**Comment by jennifer-shehane** [comment]
New PR to follow is here: https://github.com/cypress-io/cypress/pull/7154
